miden-lib/asm/sat/kernel.masm

```diff
@@ -110,82 +189,111 @@ end
 #! Panics if the ASSET is a fungible asset.
 #!
 #! Stack: [ASSET]
-#! Output: [has_asset]
+#! Output: [has_asset, 0, 0, 0]
 #!
 #! - ASSET is the non-fungible asset of interest
 #! - has_asset is a boolean indicating whether the account vault has the asset of interest
 export.account_vault_has_non_fungible_asset
+    push.0 movdn.4 push.0 movdn.4 push.0 movdn.4
+    # => [ASSET, 0, 0, 0]
+
     exec.account_vault::has_non_fungible_asset
+    # => [has_asset, 0, 0, 0]
 end
 
 #! Writes the assets of the currently executing note into memory starting at the specified address.
 #!
-#! Inputs: [dest_ptr]
+#! Inputs: [dest_ptr, 0]
 #! Outputs: [num_assets, dest_ptr]
 #!
 #! - dest_ptr is the memory address to write the assets.
 #! - num_assets is the number of assets in the currently executing note.
 export.get_note_assets
     exec.note::get_assets
+    # => [num_assets, dest_ptr, 0]
+
+    movup.2 drop
+    # => [num_assets, dest_ptr]
 end
```

Comment:
Why can't we pass memory pointers be passed across the context boundary?  

I think the way this has been implemented allows for the pattern you describe.  Internally we have, `get_vault_data` and `get_assets` procedures. We can make `get_vault_data` a kernel procedure and then use `get_vault_data` to "unhash" the vault data in the note context.  We have a #TODO inline for this - see below:

https://github.com/0xPolygonMiden/miden-base/blob/f1875705631d7839629099a0c1cec4316d750ec6/miden-lib/asm/sat/internal/note.masm#L28-L99

How are you thinking about identifying the note context?  Notes are executed in the root context so we could assert that the caller is root but this something more explicit would be more robust in my opinion.  Any ideas?

As this is only a data accessor I think it is fine to allow it to be invoked from both note and account contexts.  One point on this, if note assets are relevant in the account procedure context but memory pointers can't be passed across boundaries then how are you thinking about the account procedure gaining access to the note assets?

Reply:
> Why can't we pass memory pointers be passed across the context boundary?

We can return them, but the caller won't have access to the memory of the callee. Basically, if the `syscall` writes to memory - it will write to memory in its context. Then, if the caller tries to read from memory, it would do so from its own context which would not contain the values written by the `syscall`.

> We can make `get_vault_data` a kernel procedure and then use `get_assets` to "unhash" the vault data in the note context.

Yes, I think this will work. To summarize:
- In the kernel, instead of `get_note_assets` we'll have `get_note_vault_data`.
- The the user-facing wrapper, we'll have `notes::get_assets` which will internally make a syscall to `get_note_vault_data` procedure, and then unhash the results in the callers memory (we can have a TODO for this now).

Does this look correct?

> How are you thinking about identifying the note context?

I think we can check if the `caller` is the script root of the currently executing note. But again, I'm not sure we should impose this restriction any more.

> if note assets are relevant in the account procedure context but memory pointers can't be passed across boundaries then how are you thinking about the account procedure gaining access to the note assets?

Originally, I was thinking that notes would have to pass assets to the account via w/e interface method the account exposes. For example, for a basic wallet account interface, there would be `receive_asset` procedure which would add a single asset to the account's vault. This procedure would then be invoked from a note script.

Reply:
> Yes, I think this will work. To summarize:
> 
> * In the kernel, instead of `get_note_assets` we'll have `get_note_vault_data`.
> * The the user-facing wrapper, we'll have `notes::get_assets` which will internally make a syscall to `get_note_vault_data` procedure, and then unhash the results in the callers memory (we can have a TODO for this now).
> 
> Does this look correct?

Yes this looks correct to me.

> Originally, I was thinking that notes would have to pass assets to the account via w/e interface method the account exposes. For example, for a basic wallet account interface, there would be `receive_asset` procedure which would add a single asset to the account's vault. This procedure would then be invoked from a note script.

Got it. If we relax the constraint then we broaden the potential of account procedures.  For example we can have an account procedure `absorb_note_assets` that will batch consume the assets in a note and internally can invoke the `get_asstets` method. I don't see any reason why we wouldn't want to allow this.